src/System.Linq/tests/ToHashSetTests.cs

```diff
@@ -16,10 +17,38 @@ private class CustomComparer<T> : IEqualityComparer<T>
             public int GetHashCode(T obj) => EqualityComparer<T>.Default.GetHashCode(obj);
         }
 
+        public class Record
+        {
+            public string Name;
+            public int Score;
+
+            private sealed class RecordComparer : IEqualityComparer<Record>
+            {
+                public bool Equals(Record x, Record y)
+                {
+                    if (ReferenceEquals(x, y)) return true;
+                    if (ReferenceEquals(x, null)) return false;
+                    if (ReferenceEquals(y, null)) return false;
+                    if (x.GetType() != y.GetType()) return false;
```

Comment:
There does not seem to be a reason for all of these `ReferenceEquals` and `GetType` checks because you are only using this comparer in one place where none of these branches will be followed. You can assume nothing is null and write

```cs
public bool Equals(Record x, Record y) => string.Equals(x.Name, y.Name) && x.Score == y.Score;
```